File: 12.07.21/puzzle2.R
library(purrr)

input <- scan("12.07.21/input.txt", sep = ",")
h_max <- max(input)
h_min <- min(input)
search <- h_min:h_max
cost_function <- function(p1, p2){
  map(p2, function(x){abs(sum(0:(x-p1)))}) %>% unlist %>% sum
}

map(search, cost_function, p2 = input) %>% unlist %>% min

File: 2023/12.11.2023/puzzle2.py
import numpy as np
from itertools import combinations

with open("input.txt", "r") as f:
    lines = f.read().splitlines()
lines = [list(line) for line in lines]
lines = np.array(lines)

empty_cols = np.where(np.all(lines == ".", axis=0))[0]
empty_rows = np.where(np.all(lines == ".", axis=1))[0]


def custom_manhattan(c1, c2, expansion):
    expansion1 = len(empty_rows[(empty_rows - min(c1[0],c2[0]) > 0) & (empty_rows - max(c1[0],c2[0]) < 0)])
    expansion2 = len(empty_cols[(empty_cols - min(c1[1],c2[1]) > 0) & (empty_cols - max(c1[1],c2[1]) < 0)])
    m = np.abs(c1[0]-c2[0])+(expansion-1)*(expansion1) + np.abs(c1[1]-c2[1])+(expansion-1)*(expansion2)
    return m

galaxies = np.transpose((lines=="#").nonzero()).tolist()
print(sum([custom_manhattan(c1,c2, int(1e6)) for c1, c2 in list(combinations(galaxies, 2))]))
